doc/sisiyaqd.8
.\"(c) Copyright 2003 by Erdal Mutlu
.\"(c) Sections Copyright 2003 by Erdal Mutlu
.\"All rights reserved.  The file named COPYRIGHT specifies the terms 
.\"and conditions for redistribution.
.\"
.\" $Id: sisiyaqd.8 68 2005-02-17 17:29:46Z emutlu $
.TH SISIYAQD 8 "15 January 2005"
.\" *************************** NAME *********************************
.SH NAME
sisiyaqd \- The SisIYA queue processor server.
.\" *************************** SYNOPSIS *********************************
.SH SYNOPSIS
.B sisiyaqd
config_file_name
.\" *************************** DESCRIPTION *********************************
.SH DESCRIPTION
\fBsisiyaqd\fP is a part of SisIYA - a system monitoring tool.
\fBsisiyaqd\fP updates the SisIYA database from records of the
systemhistorystatusqueue table. This table is populated by the 
\fBsisiyad\fP server, when it recieves client messages. The
\fBsisiyaqd\fP checks periodically for new records and updates the
database. After every check it sleeps for number of seconds specified in the
SLEEP_SOCONDS parameter. The default values is 30.
.\" *********************** CONTROLLING sisiyaqd ****************************
.SH "CONTROLLING SISIYAQD"
.LP
\fBsisiyaqd\fP performs certain actions when it receives certain signals.
.TP 15
.B SIGHUP
causes a hard reconfiguration, which means that \fBsisiyaqd\fP re-reads 
the configuration file.
.TP
.B SIGQUIT
causes clean program termination.
.TP
.B SIGTERM
the same as SIGQUIT.
.TP
.B SIGUSR1
causes the \fBsisiyaqd\fP server process to print its configuration values via syslog.
.\" *********************** FILES ****************************
.SH FILES
.LP
.PD .1v
.TP 20
.B /etc/sisiyaqd.conf
default configuration file
.PD
.\" *********************** SEE ALSO ****************************
.SH "SEE ALSO"
.I "sisiyaqd.conf(5), sisiyad(8)"
.LP
.I "http://sisiya.sourcefourge.net"
.\" *********************** AUTHOR ****************************
.SH AUTHOR
Erdal Mutlu
.\" *********************** PRONUNCIATION ****************************
.SH PRONUNCIATION
sisiya



ChangeLog
0.5:
Release 1
- sisiya client check scripts on Linux/UNIX systems have now a lock file to prevent running it when the previous script is not finished
- modified sisiya_isuptodate to be used with yum, zypper and pacman
- modified sisiya_acpi_temperature check to use acpi command when the /proc/acpi/thermal_zone directory does not exist
- modified sisiya_acpi_battery check to use acpi command when the /proc/acpi/battery directory does not exist
- added isuptodate check for Linux yum based systems
- added server checks for vmware systems.
- added support for configuring remote checks from the web interface. 
- added auto update support for MS Windows client checks.
- changed the SisIYA message format from a character seperated string to XML message.
- change the statusid's to be able to add truble ticket like system to the SisIYA, which is going to be integrated soon. 
- added support for adding and/or replacing system images and changed the structure of the images folder.
- added network discovery support for network device view.
- added network device view to web GUI.
- added dashboard to the web GUI.
- added autodiscovery support.
- rewrite of the web GUI. Instead of a single php scirpt, which does everything, now we have banch of php scripts doing specific tasks.
0.4:
Release 29
- added server checks for sensors (temperature, humidity etc).
- added check for LSI Logic MegaRAID controllers used Dell servers using megacli utility.
- added CA BrightStore check for media pool scratch set on MS Windows systems.
- added SisIYA client checks for ACPI checking battery and temperature for MS Windows and Linux systems.
- added SisIYA_Client_GUI: a taskbar icon program, which shows the latest results for a Windows system.
- addapted SisIYA common client checks for OpenBSD, NetBSD, FreeBSD and openSolaris
- added ntpstat check for MS Windows systems.
- added MS Exchange mail queue check.
- added MS Exchange MAPI connectivity check.
Release 28
- added MS Exchange service health check.
- added sisiya_url to the sisiyaconf.php and updated sisiya_rss.php to use this variable.
- added CA BrightStore check for backup jobs using the ca_qmgr utility.
- added UPS check for CS121 SNMP web adapter equiped UPS systems.
- added sisiya_system_info_prtconf.sh script to get server info on UNIX systems with prtconf. 
- added sisiya_system_info_hpasm.sh script to get server info on HP systems with hpasmcli. 
- made some minor modifications on the client check scripts to run on AIX.
- added SisIYA system name to the imap_systems.conf file
- added SisIYA system name to the smtp_systems.conf file
- added SisIYA system name to the switch_systems.conf file
- added RAID level, stripe size etc info in the check for HP Smart Array controllers using hpacucli utility for MS Windows. 
- added check for the running status of MS Windows services, which are set be auto started.
- added NSIS installer script for MS Windows systems.
- added antivirus status check for MS Windows systems.
- added check for programs that must be always running for MS Windows systems.
- added load check for MS Windows systems.
- added swap check for MS Windows systems.
- added uptodate check for MS Windows systems.
- added file system check for MS Windows systems.
- added event log check for MS Windows systems.
- added system check for MS Windows systems.
- added SisIYA client checks for MS Windows systems, which use powershell.
- removed the not used host_name parameter from sisiya_send_message2.sh script.
- added check for VMware.
- added SisIYASendMessage.java: a program for sending messages to the SisIYA server.
- changed serviceid_smbd to serviceid_smb in the sisiya_check_smb_system.sh script.
- added filesystem type parameter to the file system check script.
- added RAID level, stripe size etc info in the check for HP Smart Array controllers using hpacucli utility. 
- fixed a bug in sisiyac1.c, which truncated the message's last character.
- added check for LSI Logic MegaRAID controllers using megarc utility.
- added CPU status check script for HP systems using hpasmcli.
- added RAM status check script for HP systems using hpasmcli.
- added SisIYA system name to printer_systems.conf file.
Release 26
- added check for HP Smart Array controllers using hpacucli utility.
- added check script for HP ProLiant servers with Integrated Lights-Out 2 module, which checks for fan speeds, temperatures etc using hpasmcli utility.
- improved ntpstat script when the system does not have the ntpstat command and the ntpq does not return proper return codes.
- added script to check number of mails in a mail queue.
- added script to check software RAID devices.
- added script to check EDI on Baan (Infor).
- fixed the wget timeout problem by adding --tries=1 option in sisiya_check_http_system.sh
- added SisIYA system name to the telnet_systems.conf file
Release 24
- added SNMP trap support. A script, which gets called from the snmptrapd and sends messages to SisIYA.
- added sisiya check script for monitoring online Baan users. 
- changed the SisIYA's administration web interface.
- changed language support file format, so that every language has its own file.
- changed DOCTYPE from HTML 4.01 Transitional to XHTML 1.0 Strict
- added stylesheet for the web interface
- added Baan jobs script. It checks the status of the Baan jobs.
- added SMART check for SMART enabled disks. At the moment it checks the disk's temperature and displays disk infos.
- added default Apache configuration file sisiya.conf
- changed all bash source command to "." (dot) command.
- changed all == in test to = in the if and while tests
- added user and password support for sisiya_check_http script
- added ntpq command to sisiya_ntpstat check script
- added a PHP script to generate an XML file for client applications
- added onerror option to sisiya_listening_socket check
- added a server check for DNS servers

src/sisiyaqd.c
/*
    Copyright (C) 2003  Erdal Mutlu

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifdef HAVE_CONFIG_H
 #include"config.h"
#endif

#ifdef HAVE_MYSQL
 #include"sisiyad_mysql.h"
#endif
#ifdef HAVE_PGSQL
 #include"sisiyad_postgresql.h"
#endif

#include"sisiya.h"
#include"sisiya_conf.h"

#ifdef HAVE_SYS_STAT_H
 #include<sys/stat.h>
#endif

#ifdef HAVE_SYS_TYPES_H
 #include<sys/types.h>
#endif

#ifdef SERVER
 #include"esignal.h"
#endif

#ifdef HAVE_SIGNAL_H 
 #include<signal.h>
#endif

pid_t server_pid; /* This is the PID of the daemon process. */
char *conf_file_name;
int loglevel;

char configs[NUMCONFIGS][2][MAX_STR]={
 {"LOGLEVEL","0"},
 {"IP","any"},
 {"PORT","8888"},
 {"DB_TYPE","PostgreSQL"},
 {"DB_SERVER","localhost"},
 {"DB_NAME","sisiya"},
 {"DB_USER","sisiyauser"},
 {"DB_PASSWORD","sisiyauser1"},
 {"SEMKEY","11"},
 {"PID_FILE","/var/run/sisiyad.pid"},
 {"QPID_FILE","/var/run/sisiyaqd.pid"},
 {"SLEEP_SECONDS","30"}
};

/***********************************************************************/
RETSIGTYPE sig_quit(int signo);
RETSIGTYPE sig_hup(int signo);
RETSIGTYPE sig_term(int signo);
RETSIGTYPE sig_usr1(int signo);
int daemon_init(void);
char *conf_file_name;
void clean_and_exit(int code);
void doit(void);
/***********************************************************************/
/*
 sig_quit : Signal handler for SIGQUIT.
*/
RETSIGTYPE sig_quit(int signo)
{
 if(loglevel > 1)
   syslog(LOG_INFO,"caught SIGQUIT (%d)",signo);
 clean_and_exit(EXIT_SUCCESS);
};

/*
 sig_hup : Signal handler for SIGHUP. This way we reconfigure our server.
*/
RETSIGTYPE sig_hup(int signo)
{
 syslog(LOG_INFO,"caught SIGHUP (%d). Reloading configuration.",signo);
 if(read_conf(conf_file_name) == FALSE) {
   syslog(LOG_ERR,"error occured while reading configuration file %s !",conf_file_name);
   clean_and_exit(EXIT_FAILURE);
 }
 doit();
};

/*
 sig_term : Signal handler for SIGTERM.
*/
RETSIGTYPE sig_term(int signo)
{
 if(loglevel > 1)
   syslog(LOG_INFO,"caught SIGTERM (%d)",signo);
 clean_and_exit(EXIT_SUCCESS);
};

/*
  sig_usr1 : Signal handler for SIGUSR1. Log the configuration options via syslog.
*/
RETSIGTYPE sig_usr1(int signo)
{
 sisiya_showconf();
}



/* 
  clean_and_exit: Performs clean up and exits with the specified exit code.
*/
void clean_and_exit(int code)
{
 syslog(LOG_INFO,"exiting...");
 exit(code);
}

/*
 daemon_init: Initialize a daemon process.
*/
int daemon_init(void)
{
 pid_t pid;

 /* There is a problem when I send the SIGHUP for the second time. The first SIGHUP
    is caught, but the second etc. are not. */

 pid=fork();
 if(pid < 0)
   return(-1);
 else if(pid != 0)
   exit(0); /* the parent exits */

 /* the child process continues */
 pid=setsid(); /* become session leader */  
 if(pid == -1) {
  syslog(LOG_ERR,"cannot set session id");
  exit(1);
 }
 
/* change working dir to /, so that we do not occupie any mounts */ 
 if(chdir("/") != 0) {
  syslog(LOG_ERR,"cannot change dir to /");
  exit(1);
 }
 
 umask(0); /* clear our file creation mask, so that any inherited restriction do not apply */ 
 return(0);
};

void doit(void)
{
	/* become a daemon */
	daemon_init(); 
 
signal(SIGCHLD,SIG_IGN);
	signal(SIGQUIT,sig_quit);
	signal(SIGTERM,sig_term);
	signal(SIGUSR1,sig_usr1);
	signal(SIGHUP,sig_hup);

	server_pid=getpid();
	if(write_pid(configs[QPID_FILE][1],server_pid) == FALSE) {
		syslog(LOG_ERR,"can't write PID to the file : %s ! Exiting...",configs[QPID_FILE][1]);
		clean_and_exit(EXIT_FAILURE); 
	}
 
	loglevel=atoi(configs[LOGLEVEL][1]); /* Set the loglevel. For now it is primitive made. Change it later.*/
	syslog(LOG_INFO,"Queue processor server (version %s) started. PID is %d",VERSION,server_pid); 

	for(;;) {
#ifdef HAVE_MYSQL 
		if(strcmp(configs[DB_TYPE][1],"MySQL") == 0)
			sisiyad_mysql_process_queue(configs[DB_SERVER][1],configs[DB_USER][1],configs[DB_PASSWORD][1],configs[DB_NAME][1]);
		else
			syslog(LOG_ERR,"This type [%s] is not MySQL. The queue is not processed!",configs[DB_TYPE][1]);

#else
 #ifdef HAVE_PGSQL
		if(strcmp(configs[DB_TYPE][1],"PostgreSQL") == 0)
			sisiyad_postgresql_process_queue(configs[DB_SERVER][1],configs[DB_USER][1],configs[DB_PASSWORD][1],configs[DB_NAME][1]);
		else
			syslog(LOG_ERR,"This type [%s] is not PostgreSQL. The queue is not processed!",configs[DB_TYPE][1]);

 #else
		syslog(LOG_ERR,"unknown DB_TYPE [%s]. Queue table is not processed!",configs[DB_TYPE][1]);
		return;
 #endif
#endif
		/* sleep for while */
		sleep(atoi(configs[SLEEP_SECONDS][1]));	
	}
};



int main(int argc,char **argv)
{

 if(argc != 2 ) {
  fprintf(stderr,"Usage : %s sisiyaqd.conf\n",argv[0]);
  exit(1);
 }

 openlog("sisiyaqd",LOG_PID,LOG_USER); /* Open the syslog. Log progname and pid */

 conf_file_name=argv[1];
 if(read_conf(conf_file_name) == FALSE) {
   syslog(LOG_ERR,"error occured while reading configuration file %s !",conf_file_name);
   clean_and_exit(EXIT_FAILURE);
 }

 /* everything goes here */
 doit();
 
 return(0);
}
